src/librustc_typeck/check/generator_interior.rs

```diff
@@ -172,21 +172,27 @@ pub fn resolve_interior<'a, 'tcx>(
     });
 
     // Store the generator types and spans into the tables for this generator.
-    let types = types
+    let mut captured_tys = FxHashSet::default();
+    let type_causes = types
         .into_iter()
         .zip(&folded_types)
-        .map(|((mut interior_cause, _), ty)| {
-            interior_cause.ty = ty;
-            interior_cause
+        .filter_map(|((mut interior_cause, _), ty)| {
+            let ty = fcx.tcx.erase_regions(ty);
```

Comment:
`erase_regions` doesn't actually erase all regions. You'll want another `fold_regions` call where the closure returns `tcx.lifetimes.re_erased`. It would also make sense to deduplicate before we replace the erased regions with late bound regions.